src/core/fetchers.ts

```diff
@@ -0,0 +1,151 @@
+import { isExperienceEntry } from '../typeguards'
+
+type fetchExperienceEntitiesArgs = {
+  client: ContentfulClientApi<undefined>
+  experienceEntry: Entry
+  locale: string
+}
+
+export const fetchExperienceEntities = async ({
+  client,
+  experienceEntry,
+  locale,
+}: fetchExperienceEntitiesArgs) => {
+  if (!client) {
+    throw new Error(
+      'Failed to fetch experience entities. Required "client" parameter was not provided'
+    )
+  }
+
+  if (!locale) {
+    throw new Error(
+      'Failed to fetch experience entities. Required "locale" parameter was not provided'
+    )
+  }
+
+  if (!isExperienceEntry(experienceEntry)) {
+    throw new Error(
+      'Failed to fetch experience entities. Provided "experienceEntry" does not match experience entry schema'
+    )
+  }
+
+  const entryIds: string[] = []
+  const assetIds: string[] = []
+
+  for (const dataBinding of Object.values(experienceEntry.fields.dataSource)) {
+    if (!('sys' in dataBinding)) {
+      continue
+    }
+    if (dataBinding.sys.linkType === 'Entry') {
+      entryIds.push(dataBinding.sys.id)
+    }
+    if (dataBinding.sys.linkType === 'Asset') {
+      assetIds.push(dataBinding.sys.id)
+    }
+  }
+
+  const [entriesResponse, assetsResponse] = await Promise.all([
+    entryIds.length > 0 ? client.getEntries({ 'sys.id[in]': entryIds, locale }) : { items: [] },
+    assetIds.length > 0 ? client.getAssets({ 'sys.id[in]': assetIds, locale }) : { items: [] },
+  ])
+
+  return {
+    entries: (entriesResponse.items ?? []) as Entry[],
+    assets: (assetsResponse.items ?? []) as Asset[],
+  }
+}
+
+type fetchExperienceEntryArgs = {
+  client: ContentfulClientApi<undefined>
+  experienceTypeId: string
+  locale: string
+  idenifier: {
+    slug?: string
+    id?: string
+  }
+}
+
+export const fetchExperienceEntry = async ({
+  client,
+  experienceTypeId,
+  locale,
+  idenifier,
+}: fetchExperienceEntryArgs) => {
+  if (!client) {
+    throw new Error(
+      'Failed to fetch experience entities. Required "client" parameter was not provided'
+    )
+  }
+
+  if (!locale) {
+    throw new Error(
+      'Failed to fetch experience entities. Required "locale" parameter was not provided'
+    )
+  }
+
+  if (!experienceTypeId) {
+    throw new Error(
+      'Failed to fetch experience entities. Required "experienceTypeId" parameter was not provided'
+    )
+  }
+
+  if (!idenifier.slug && !idenifier.id) {
+    throw new Error(
+      `Failed to fetch experience entities. At least one identifier must be provided. Received: ${JSON.stringify(
+        idenifier
+      )}`
+    )
+  }
+
+  const filter = idenifier.slug ? { 'fields.slug': idenifier.slug } : { 'sys.id': idenifier.id }
+
+  const entries = await client.getEntries({
+    content_type: experienceTypeId,
+    locale,
+    ...filter,
+  })
+
+  if (entries.items.length > 1) {
+    throw new Error(
+      `More than one experience with identifier: ${JSON.stringify(idenifier)} was found`
+    )
+  }
```

Comment:
Did we have this behavior before? I thought that before we decided to take the first one if there were multiple ones. Not saying that I'm in favour of any of both ways but just wondering whether we consider this a change we need to document.

Reply:
yes, we did. Before this PR, it was part of the `useExperienceBuilderStore` hook.

Also we export these functions, so it makes sense in my opinion to make it atomic, strict and avoid assumptions. These are meant to be just helper functions to save a couple of lines of code. If one of them doesn't fit the use case the devs might have, then they can also fetch the entity in another way, which they prefer most

Reply:
You can see [the code here](https://github.com/contentful/experience-builder/blob/main/src/hooks/useExperienceStore.ts#L127-L134)